quic/s2n-quic-transport/src/stream/controller.rs

```diff
@@ -69,122 +66,181 @@ impl Controller {
     ) -> Self {
         Self {
             local_endpoint_type,
-            bidi_controller: ControllerPair {
-                stream_id: StreamId::initial(local_endpoint_type, StreamType::Bidirectional),
-                local_initiated: LocalInitiated::new(
-                    initial_peer_limits.max_streams_bidi,
-                    initial_local_limits.max_streams_bidi,
-                ),
-                remote_initiated: RemoteInitiated::new(initial_local_limits.max_streams_bidi),
-            },
-            uni_controller: ControllerPair {
-                stream_id: StreamId::initial(local_endpoint_type, StreamType::Unidirectional),
-                local_initiated: LocalInitiated::new(
-                    initial_peer_limits.max_streams_uni,
-                    stream_limits
-                        .max_open_local_unidirectional_streams
-                        .as_varint(),
-                ),
-                remote_initiated: RemoteInitiated::new(initial_local_limits.max_streams_uni),
-            },
+            local_bidi_controller: LocalInitiated::new(
+                initial_peer_limits.max_streams_bidi,
+                stream_limits.max_open_local_bidirectional_streams,
+            ),
+            remote_bidi_controller: RemoteInitiated::new(initial_local_limits.max_streams_bidi),
+            local_uni_controller: LocalInitiated::new(
+                initial_peer_limits.max_streams_uni,
+                stream_limits.max_open_local_unidirectional_streams,
+            ),
+            remote_uni_controller: RemoteInitiated::new(initial_local_limits.max_streams_uni),
         }
     }
 
     /// This method is called when a `MAX_STREAMS` frame is received,
     /// which signals an increase in the available streams budget.
     pub fn on_max_streams(&mut self, frame: &MaxStreams) {
         match frame.stream_type {
-            StreamType::Bidirectional => self.bidi_controller.local_initiated.on_max_streams(frame),
-            StreamType::Unidirectional => self.uni_controller.local_initiated.on_max_streams(frame),
+            StreamType::Bidirectional => self.local_bidi_controller.on_max_streams(frame),
+            StreamType::Unidirectional => self.local_uni_controller.on_max_streams(frame),
         }
     }
 
     /// This method is called when the local application wishes to open a new stream.
     ///
+    /// This API requires that only one stream is opened per invocation and must be
+    /// called the next stream id of a type.
```

Comment:
not sure I understand the second part of this comment "must be called the next stream id of a type."

Reply:
This api expects and only checks if a single stream can be opened. Only the type is used from the stream_id. I think its an artifact of the old API and i think we should only pass in the type to make it more clear.

Reply:
updated api and clarified the docs in https://github.com/aws/s2n-quic/pull/1406/commits/425e7183fa391bdf36dce565df5795edf6c25c2c

By only taking StreamType and not an Id, we wont expect to be able to open an arbitrary stream.. but only the next one.